{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMFw5FmeUi6xlPv07g1D3uJ"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["**Libraries**"],"metadata":{"id":"-KBL4_ZDJE55"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"b7QWmu6uJBds"},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","from scipy.stats import ttest_ind\n","import matplotlib.pyplot as plt\n","import time"]},{"cell_type":"markdown","source":["**Read the snRNA-seq sort by cell type (Ej. GABA_1, GABA_1....Gluata_1, Gluta_2....Non-neuronal_1....) or by Class**\n","\n","**The rows are genes and the columns cells**"],"metadata":{"id":"CHpRaQrMJMMm"}},{"cell_type":"code","source":["sort=pd.read_csv(\"gene_expression_matrix.csv\", index_col=0, nrows=10)\n","print(\"\\n-----------------------------------------------------------------------\\n\")\n","print(\"\\n Gene expression matrix \\n\")\n","print(sort)"],"metadata":{"id":"f3d6sJSSJJ0w"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**In this case the snRNA-seq array has phenotype (clusters) names, a file with the range of phenotypes belonging to the main class (e.g. GABAergic, Glutamatergic, Non-neuronal) is needed**\n"],"metadata":{"id":"gctujZP3JeqL"}},{"cell_type":"code","source":["range_types=pd.read_csv(\"cluster_labels_sort.csv\")\n","print(\"\\n-----------------------------------------------------------------------\\n\")\n","print(\"\\n Cluster ranges \\n\")\n","print(range_types)"],"metadata":{"id":"M2MXEUvsJcgC"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**The names of the class (columsn on the snRNA-seq)**\n"],"metadata":{"id":"eSTQ9FfCJmyd"}},{"cell_type":"code","source":["# Name of each file\n","\n","iterar=[\n","    \"Exc\",\n","    \"Inh\",\n","    \"Non-neuronal\"\n","]"],"metadata":{"id":"pSxvWOfOJlGV"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**p_value and Fold Change parameters for each gene in the each group (GABAergic, Glutamatergic, Non-neuronal)**\n"],"metadata":{"id":"b0hGr8jPJwsK"}},{"cell_type":"code","source":["for phenotype in iterar:\n","\n","        # In this section we sort each cell belonging to main class (GABAergic, Glutamatergic, Non-neuronal) in the first iteration\n","        # the \"cell\" variable contain all cell types that belonging to class of interest to have the parameters (p_value and Fold), and the \"noncell\"\n","        # variable has the remaining cells. In this first iteration we obtain the parameters for each gene in the class of interest, and this cycle will repeat\n","        # this cycle will be repeated until all the groups are completed (the main cell classes).\n","\n","\n","    # Obtain indexes to use as ranges\n","    cell_type=range_types[\"Cell type\"]==phenotype\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"\\n For each phenotype we obtain the TRUE \\n\" ) #Check point\n","    print(cell_type)\n","\n","\n","    # Create range for specific cell type \"group1\" change cell type of interest\n","    cell=range_types.loc[range_types[\"Cell type\"]==phenotype]\n","    cell=cell[\"Unnamed: 0\"].to_list()\n","    # Range of phenotype of interest\n","    print(cell)\n","\n","    # Create range for the other cell types \"group2\" change cell type of interest\n","    noncell=range_types.loc[range_types[\"Cell type\"]!=phenotype]\n","    noncell=noncell[\"Unnamed: 0\"].to_list()\n","    # Range of the rest of the phenotypes\n","    print(noncell)\n","\n","\n","\n","        # First step, separate the matrix into comparison groups.\n","\n","    # Define two groups of samples (e.g. Neruon specific type vs. all others)\n","    # Depending on the ranks obtained above, the ranks to be taken for each of the groups will be as follows\n","\n","    group1 = sort.columns[cell]  # Phenotype to compare with all others 0-xxx\n","    group2 = sort.columns[noncell] # Rest of phenotypes xxx-xxx-\n","\n","    # Data visualization\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"\\n View the groups \\n\" ) #Check point\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"Neruon specific type \\n\")\n","    print(group1)\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"Others \\n\")\n","    print(group2)\n","\n","\n","\n","        # Second step\n","\n","    # Calculate the p-value for each gene using a t-test\n","\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"\\n P_values \\n\" ) #Check point\n","    p_values = []\n","    #t_statitic= []\n","    # For each gene, the data for that gene is compared with all the columns\n","    for gene in sort.index:\n","        t, p = ttest_ind(sort.loc[gene, group1], sort.loc[gene, group2], equal_var=False)\n","        p_values.append(p)\n","\n","        print(\"\\n-----------------------------------------------------------------------\\n\")\n","        print(\"View p_value in each group\",sort.loc[gene, group1])\n","        sort.loc[gene, group1].to_csv(\"Dataframe_group1.csv\")\n","        sort.loc[gene, group2].to_csv(\"Dataframe_group2.csv\")\n","\n","\n","        #t_statitic.append(t)\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"P_values \\n\" ) #Check point\n","    print(p_values)\n","\n","\n","    sort[\"p_values\"]=p_values\n","    #sort[\"t_statitic\"]=t_statitic\n","\n","\n","        #Third step\n","\n","    # Calculate the fold change for each gene\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"Fold change \\n\" ) #Check point\n","    fold_changes = (sort[group1].mean(axis=1) / sort[group2].mean(axis=1)).apply(np.log2)\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"Fold change \\n\")\n","    print(fold_changes)\n","\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"\\n Fold change_grupo2 \\n\")\n","    print(sort[group2].mean(axis=1))\n","\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"\\n Fold change_grupo1 \\n\")\n","    print(sort[group1].mean(axis=1))\n","\n","\n","    # Add the fold changes to the gene expression matrix\n","    sort['fold_change'] = fold_changes\n","    #print(sort)\n","\n","\n","    # Save datafrmar with parameter for each gene of the class of interest\n","    sort[[\"p_values\",\"fold_change\"]].to_csv(phenotype+\"_pvalue_foldchange.csv\")\n","\n","    print(\"\\n-----------------------------------------------------------------------\\n\")\n","    print(\"The process ended in: \" +phenotype)\n","    #time.sleep(60)\n","\n","print(\"\\n-----------------------------------------------------------------------\\n\")\n","print(\"Finished\")\n","#time.sleep(60)\n"],"metadata":{"id":"5G3D71mFJzeW"},"execution_count":null,"outputs":[]}]}